src/cpu/flexcpu/flexcpu_thread.cc

```diff
@@ -569,11 +605,73 @@ FlexCPUThread::markFault(shared_ptr<InflightInst> inst_ptr, Fault fault)
     commitAllAvailableInstructions();
 }
 
+void
+FlexCPUThread::onBranchPredictorAccessed(std::weak_ptr<InflightInst> inst,
+                                         BPredUnit* pred)
+{
+    const shared_ptr<InflightInst> inst_ptr = inst.lock();
+    if (!inst_ptr
+     || inst_ptr->isCommitted()
+     || inst_ptr->isSquashed()
+     || inst_ptr->isComplete()) {
+        freeBranchPredDepth();
+        // No need to do anything for an instruction that has been
+        // squashed, or a branch that has already been resolved.
+        return;
+    }
+
+    // We check before requesting one from the CPU, so this function should
+    // only be called with a non-NULL branch predictor.
+    panic_if(!pred, "I wasn't programmed to understand nullptr branch "
+                    "predictors!");
+
+    TheISA::PCState pc = inst_ptr->pcState();
+
+    const bool taken = pred->predict(inst_ptr->staticInst(),
+                                     inst_ptr->issueSeqNum(), pc, threadId());
+
+    // BPredUnit::predict takes pc by reference, and updates it in-place.
+
+    DPRINTF(FlexCPUBranchPred, "(seq %d) predicted %s (predicted pc: %#x).\n",
+                              inst_ptr->seqNum(),
+                              taken ? "taken" : "not taken",
+                              pc.instAddr());
+    // Count stats using the predict return value?
+
+    StaticInstPtr static_inst = inst_ptr->staticInst();
+    if (static_inst->isDirectCtrl() && static_inst->isUncondCtrl()) {
```

Comment:
Is the `&&` right here? It shouldn't be `||`? just checking...

Reply:
Pretty sure it's right, since indirectctrl can't be known early as "taken" or "not taken", since a register value used to jump to could be the next instruction, whereas this logic describes an assumption that a compiler wouldn't create an unconditional direct jump to the immediate next instruction.